### PR TITLE
feat: image components!

### PR DESCRIPTION
**Pull request by jennspencer**
[![PR App][icn]][demo] | Fix RM-9020
:-------------------:|:----------:

## 🧰 Changes

MDXify the `Image` component

## 🧬 QA & Testing

- [Broken on production][prod].
- [Working in this PR app][demo].

[demo]: https://markdown-pr-PR_NUMBER.herokuapp.com
[prod]: https://SUBDOMAIN.readme.io
[icn]: https://user-images.githubusercontent.com/886627/160426047-1bee9488-305a-4145-bb2b-09d8b757d38a.svg

### CODE REVIEW

File: processor/compile/image.js [hunk 1]
```diff
@@ -7,11 +7,6 @@ module.exports = function ImageCompiler() {
   visitors.image = function compile(node, ...args) {
     if (node.data?.hProperties?.className === 'emoji') return node.title;
 
-    // Use magic block instead of markdown when there are certain defined properties we can't store in markdown
-    const { align, className, width } = node.data?.hProperties || {};
-    const useMagicBlock = Boolean(width) || className?.length || Boolean(align);
-    if (useMagicBlock) return visitors.figure.call(this, node, ...args);
```
Comment (kellyjosephprice) [inline processor/compile/image.js:13]: I _think_ we'll need to replace this something that writes out a jsx tag, right? Something like:

```
return `<Image width="${width}" />`
```

**Review by kellyjosephprice — CHANGES_REQUESTED** [review]
This looks good. But I think we need to replace the magic block stuff with some equivalent MDX?

Also, would you try enabling tests in these files:
```
__tests__/components/test.js
__tests__/flavored-compilers/images.test.js
__tests__/index.test.js
```

File: processor/compile/image.js (hunk 1 above)
Comment (jennspencer) [reply processor/compile/image.js:13]: it didn't seem to be an issue? we're not having to store the extra properties in a magic block anymore, so the component just spits out the correct JSX appropriately (unless I've gotten all this wrong, which is also entirely possible 😂 )

Comment (kellyjosephprice) [reply processor/compile/image.js:13]: The compilers take an `ast` and spit out `md/mdx`. This mostly happens when the editor saves, so I'm pretty sure saving images with options won't save the extra options until we replace the magic block stuff.

Comment (jennspencer) [reply processor/compile/image.js:13]: how do i test this? are we able to use the RDMX flag in the main app yet?

Comment (kellyjosephprice) [reply processor/compile/image.js:13]: This [file](https://github.com/readmeio/markdown/blob/next/__tests__/flavored-compilers/images.test.js#L4) really should have had a magic block test.

And yea, the flag ~works. A lot of things can crash the page.

File: processor/compile/image.js [hunk 2]
```diff
@@ -7,10 +7,9 @@ module.exports = function ImageCompiler() {
   visitors.image = function compile(node, ...args) {
     if (node.data?.hProperties?.className === 'emoji') return node.title;
 
-    // Use magic block instead of markdown when there are certain defined properties we can't store in markdown
     const { align, className, width } = node.data?.hProperties || {};
-    const useMagicBlock = Boolean(width) || className?.length || Boolean(align);
-    if (useMagicBlock) return visitors.figure.call(this, node, ...args);
+    const complexImage = Boolean(width) || className?.length || Boolean(align);
+    if (complexImage) return <Image {...node.data.hProperties} />;
```
Comment (kellyjosephprice) [inline processor/compile/image.js:12]: This needs to return a string?

Comment (jennspencer) [reply processor/compile/image.js:12]: ahh, so like a template literal or `renderToString`?

honestly i'm not quite sure what this step is doing, so i'm kinda winging it haha

Comment (kellyjosephprice) [reply processor/compile/image.js:12]: Just a template literal. This is when the editor saves back to markdown/mdx. The editor calls `serialize`/`toString` which converts the slate value to `mdast`. Then it passes that to `RMDX.mdx`, which converts the `mdast` to a plain text string using all the compilers.

**Comment by jennspencer** [comment]
@kellyjosephprice okay that test is being a pain and we can address it later :P

File: processor/compile/image.ts [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+
+  visitors.image = function compile(node: Image, ...args: any[]): string {
+    if (node.data?.hProperties?.className === 'emoji') return node.title;
+
+    const { align, className, width } = node.data?.hProperties || {};
+    const complexImage: boolean = Boolean(width) || Boolean(className) || Boolean(align);
+    if (complexImage) return `<Image ${{...node.data?.hProperties}} />`;
+
+    return originalImageCompiler.call(this, node, ...args);
+  };
```
Comment (kellyjosephprice) [inline processor/compile/image.ts:17]: This is the old (`remark-stringify@8`) syntax. Here's an example of the a new-style plugin: https://github.com/readmeio/markdown/blob/beta/processor/compile/gemoji.ts#L1-L13

Also, my code-tabs PR refactors it a little: https://github.com/readmeio/markdown/pull/871/files#diff-3dd941d5e7f0c5b195b587d965c70f0e20f53da74c480b7d208e555b0dca2715R1-R12

File: package.json [hunk 4]
```diff
@@ -19,6 +19,7 @@
     "release.dry": "npx semantic-release --dry-run",
     "start": "webpack serve --open --mode development --config ./webpack.dev.js",
     "test": "vitest",
+    "test.ui": "vitest --ui",
```
Comment (kellyjosephprice) [inline package.json:22]: :neat:

**Review by kellyjosephprice — APPROVED** [review]
Mostly LGTM, accept for the compiler's using the old style syntax.

File: processor/compile/image.ts (hunk 3 above)
Comment (kellyjosephprice) [reply processor/compile/image.ts:17]: I guess I wasn't aware of this 3 weeks ago...

Comment (jennspencer) [reply processor/compile/image.ts:17]: hah! awesome--i had seen the syntax change but wasn't sure if that was just gemoji related

Comment (kellyjosephprice) [reply processor/compile/image.ts:17]: It's like, well typed, but I could only figure it out by reading the remark source. 😖

File: tsconfig.json [hunk 5]
```diff
@@ -15,8 +15,21 @@
...
+    "./contexts",
+    "./example",
+    "./lib",
+    "./processor"
+  ],
+  "exclude": [
+    "node_modules",
+    "dist"
+  ]
+}
```
Comment (jennspencer) [inline tsconfig.json:35]: my json formatter has opinions apparently

**Review by kellyjosephprice — APPROVED** [review]
LGTM!

:pray: Would you uncomment this line: https://github.com/readmeio/markdown/blob/beta/__tests__/browser/markdown.test.js#L20

Then run `make updateSnapshot`

**Comment by jennspencer** [comment]
> LGTM!
> 
> 🙏 Would you uncomment this line: https://github.com/readmeio/markdown/blob/beta/__tests__/browser/markdown.test.js#L20
> 
> Then run `make updateSnapshot`

heh we have dueling default fonts

**Comment by kellyjosephprice** [comment]
> heh we have dueling default fonts

Ugh, docker was supposed to fix this.

**Comment by jennspencer** [comment]
> > heh we have dueling default fonts
> 
> Ugh, docker was supposed to fix this.

haha i know, i just tried to cheat and not run it in a container but then i gave in

**Comment by rafegoldberg** [comment]
## This PR was released!

🚀 Changes included in [<b>v6.75.0-beta.29</b>](/readmeio/markdown/releases/tag/v6.75.0-beta.29)
<br>

**Pull request merged**